### PR TITLE
Get rid of FlutterNativeView, introduce FlutterEngine and FlutterRenderer

### PR DESCRIPTION
**Pull request by matthew-carroll**
**Note: Please ignore the /legacy package in terms of architecture or code quality concerns.  The creation of that package was a required step for compilation.  I will move classes from /legacy into appropriate packages as I consider them truly "updated" for this next version of the API.**

### Overarching ideas of this PR

**Android is Android, and Flutter is Flutter**
We have had a strong entanglement of core Android framework artifacts with core Flutter artifacts.  This is especially clear in FlutterView.  This PR takes a major step forward in the process of clearly delineating the Android framework classes that happen to render Flutter stuff, vs Flutter classes that can be rendered by the Android framework.

**Breakdown Flutter classes by responsibility**
Previously, the notion of a "flutter engine" and a "flutter view" were completely fused in FlutterNativeView.  This PR introduces 2 important encapsulation boundaries.  

First, this PR sets up a composition relationship whereby a FlutterEngine contains within it a reference to a FlutterRenderer that may, or may not be attached to a RenderSurface.  This relationship makes explicit the notion of a FlutterEngine running without any UI, or in some manner existing between UIs.

Second, this PR introduces a FlutterJNI class which contains all JNI method declarations.  It's unclear if I'll keep them all in one place by the end of the refactor, but with FlutterJNI we are now in a position to introduce an Interface boundary around the JNI calls which will then allow for easy mocking for tests or any other purpose.

### PR Wins

**No more FlutterNativeView**
FlutterNativeView was a Class with a confusing name and an amalgamation of responsibilities.  That class has now been replaced by a combination of FlutterEngine, FlutterRenderer, and FlutterJNI.

**No more native calls in FlutterView**
FlutterView is an Android View class.  As such, we don't want to create any conflict for lifecycle or execution control.  Therefore, the presence of JNI methods in FlutterView was a problem.  Those methods meant that FlutterView needed to be passed around to other Flutter related Objects which then entangled Android lifecycle concerns with Flutter lifecycle concerns.

This PR has extracted all JNI methods from FlutterView (and FlutterNativeView) and placed them in FlutterJNI.

**FlutterView no longer shares a lifespan with FlutterNativeView/FlutterEngine**
Even though we say that Flutter can technically be attached to, and detached from a FlutterView, the FlutterView constructor actually required the existence of a FlutterNativeView AKA FlutterEngine.  Thus, no execution path existed to add, remove, add, remove, etc. a given FlutterNativeView AKA FlutterEngine to/from a given FlutterView.

This PR separates FlutterView construction from the idea of attaching to, and detaching from a FlutterEngine.  Now, for the first time, a FlutterView can exist without any reference to a FlutterEngine.  It is up to the developer when to attach FlutterView to a given FlutterEngine.  This looser coupling allows a view hierarchy to contain a FlutterView while the developer does other work, and then when the other work is done, the developer can simply attach a FlutterEngine to the existing FlutterView.  This is generally the paradigm for Android Views.  Consider a WebView that sits idle until the developer wants to load a web page.  Consider a ListView that sits empty until the developer wants to attach a ListAdapter.  The examples go on, but in general an Android View should be able to exist without issue even when it is not connected to any "content".   This is now true of FlutterView.

**Removed all lifecycle methods from FlutterView**
FlutterView previously replicated every Activity/Fragment lifecycle method so that those calls could be forwarded to the plugin system that was partially stuck within FlutterView.  Now, instead of FlutterFragment forwarding those calls to the FlutterView, which would then forward those calls to FlutterNativeView, the FlutterFragment holds onto a FlutterEngine and the FlutterFragment reports these lifecycle events directly.

**Added JavaDocs**
I drafted initial class JavaDocs for FlutterEngine, FlutterRenderer, RenderSurface, and FlutterView.

### Issues Introduced by this PR

**Lots of logging**
This PR added lots of log statements.  These logs help ensure that I don't break basic execution paths.  They will eventually be removed as the refactor stabilizes, and as tests are added.

**Uglier platform_view_android_jni.cc**
I'm new to JNI so I made a bit of a mess of the C binding.  I'll clean this up once I reach a final Java breakdown of JNI calls.

**New /legacy package**
Due to a high level of entanglement of Android embedding classes, in conjunction with copious use of package private modifiers, I had no choice in this PR but to copy over a number of source files.  Furthermore, as I made changes to FlutterView, FlutterNativeView, FlutterEngine, and FlutterRenderer, a number of those files needed to have reference types updated.  For example, FlutterNativeView was deleted, so anything that expected to receive a FlutterNativeView needed to be updated.

**Movement of lifecycle calls and some plugins from FlutterView to FlutterFragment**
This actually isn't a negative, but it might initially seem that way.  It might seem bad to move responsibilities up the dependency graph instead of down.  But here is what I'm thinking....

Up to this point we have failed to distinguish between Activity/Fragment lifecycles and View lifecycles.  The Flutter embedding has setup many explicit and implicit dependencies on these lifecycles as if all such lifecycles are universally available.  This isn't the case.  Maybe an Activity is available, maybe it isn't. Maybe a View is available, maybe it isn't.  In this next API iteration it is important that we find a way to place responsibility where it belongs.

Therefore, I made FlutterFragment responsible for invoking any Flutter methods that are related to the Activity/Fragment lifecycle.  When I extracted that logic from FlutterView, it required that I also move a few plugins from FlutterView to FlutterFragment.  I think this is a code smell that suggests that our current treatment of plugins within FlutterView is probably incorrect.

In a followup PR I intend to look at how we setup all default plugins within FlutterFragment and FlutterView and see if it deserves a different treatment in general.  But for now the current shift to FlutterFragment is a step in the right direction.

### CODE REVIEW

**Comment by amirh** [comment]
Are we going to lose git history for the files that you are copying to legacy?

**Comment by amirh** [comment]
Actually seems like we're going to lose history for e.g FlutterView as well, is it possible to `git mv` instead of adding copies? if that results in an intermediate broken branch maybe it's worth working in a branch until stuff is stable?

I think we should try harder to maintain git history + it should make the review easier to see what was moved as-is and what are the deltas (hopefully if Github's codreview UI does that for git mv)

**Comment by amirh** [comment]
Ohh we are on a staging branch :smile:, anyway can we do something to matntain git history, and show a reasonable diff for review? (5k lines to read is quite a lot for one PR)

File: shell/platform/android/platform_view_android_jni.cc [hunk 1]
```diff
@@ -757,52 +758,311 @@ bool PlatformViewAndroid::Register(JNIEnv* env) {
...
+
+  g_flutter_callback_info_constructor = env->GetMethodID(
+      g_flutter_callback_info_class->obj(), "<init>",
+      "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)V");
+  if (g_flutter_callback_info_constructor == nullptr) {
+    FML_LOG(ERROR) << "Could not locate FlutterCallbackInformation constructor";
+    return false;
+  }
+
+  // TODO(mattcarroll): this assumes the use of the new API is based on what we're compiling against but that's not true. Need to support both simultaneously.
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/platform_view_android_jni.cc]: Does this mean we need to update build files to generate different flutter.jar's?

At the moment this .cc file contains support for the old way of doing things, and the new one.  My TODO comment was pointing out that the compile-time existence of the new Java class doesn't imply that the developer is necessarily using the new Java class.  So the TODO says we need explicit instruction from the Java side as to which universe we're supporting.

I'm not sure off the top of my head how to support customized JARs because developers pull a single build of the engine with the framework, right?  How would a developer choose one JAR or the other in that scenario?  

Or am I totally misunderstanding your question?

**Comment by matthew-carroll** [comment]
@amirh Fair point about git history.  I don't know that we can preserve it for FlutterActivity or FlutterFragment, I'm also not sure we want to, but I'm trying to figure out how to at least pull in the history for everything in /legacy.

File: shell/platform/android/platform_view_android_jni.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jason-simmons) [reply shell/platform/android/platform_view_android_jni.cc]: Does the flutter.jar built on your branch include both the old and new Java classes?  Or does it only include the new classes?

If the former, then the JNI code should register implementations for both the old and new method signatures.  If the latter, then you can remove the registration code for the old classes.

**Comment by matthew-carroll** [comment]
Alright, after committing all copied source files in its own PR and then rebasing, the new code count went from 5,000 lines to just under 2,000.  And now there is almost 2,000 lines removed, too.

CC @amirh

File: shell/platform/android/io/flutter/embedding/FlutterFragment.java [hunk 2]
```diff
@@ -164,7 +179,7 @@ public void onBackPressed() {
    * @param grantResults permission grants or denials
    */
   public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
-    flutterView.getPluginRegistry().onRequestPermissionsResult(requestCode, permissions, grantResults);
+    flutterEngine.getPluginRegistry().onRequestPermissionsResult(requestCode, permissions, grantResults);
```
Comment (mehmetf) [inline shell/platform/android/io/flutter/embedding/FlutterFragment.java]: What is the relationship between these new entities when there are multiple FlutterViews?

Previously, it was implied that each FlutterView had one FlutterNativeView (with its own registry and thus plugins) unless you explicitly told it to share a FlutterNativeView. Does that change now?

File: shell/platform/android/io/flutter/embedding/FlutterRenderer.java [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        return;
+      }
+      unregisterTexture(id);
+      surfaceTexture.release();
+      released = true;
+    }
+  }
+  //------ END TextureRegistry IMPLEMENTATION ----
+
+  // TODO(mattcarroll): what exactly is this method intended to do?
```
Comment (mehmetf) [inline shell/platform/android/io/flutter/embedding/FlutterRenderer.java]: TMK these are methods called by Android system when the texture lifecycle changes (occluded/needs redrawing etc).

File: shell/platform/android/io/flutter/embedding/FlutterRenderer.java [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+        physicalPaddingLeft,
+        physicalViewInsetTop,
+        physicalViewInsetRight,
+        physicalViewInsetBottom,
+        physicalViewInsetLeft
+    );
+  }
+
+  // TODO(mattcarroll): why does this method exist?
+  public Bitmap getBitmap() {
```
Comment (mehmetf) [inline shell/platform/android/io/flutter/embedding/FlutterRenderer.java]: Are you doing this abstraction to make this class more testable by passing in a different implementation of FlutterJNI?

which brings me to the next point. It would be great to have tests for these to make sure they are doing the "right thing". :-).

File: shell/platform/android/io/flutter/embedding/FlutterFragment.java (hunk 2 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterFragment.java]: The idea I'm working towards is that you can think of a FlutterView as being completely independent of the Flutter engine/isolate/rendering that backs it.  As such, you can attach and detach FlutterEngines to any FlutterViews at will.

The analogy I made in the PR write-up compared FlutterView to a ListView.  You can attach and detach ListAdapters whenever you want and there is no lasting relationship between a ListView and a detached ListAdapter.

Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterFragment.java]: And, about plugins, I expect every FlutterEngine to have its own PluginRegistrar and plugin context.

File: shell/platform/android/io/flutter/embedding/FlutterRenderer.java (hunk 4 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterRenderer.java]: One goal is testability.  Another goal is facilitating a mental model of the Java implementation that doesn't require thinking about mixtures of Java and native calls.  I think it's easier to explain relationships when you don't have somewhat random jumps from Java code to C code.

File: shell/platform/android/io/flutter/embedding/FlutterRenderer.java [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+    return flutterJNI.nativeGetBitmap(nativeObjectReference);
+  }
+
+  // TODO(mattcarroll): what exactly is this method intended to do?
```
Comment (amirh) [inline shell/platform/android/io/flutter/embedding/FlutterRenderer.java]: This method passes touch events to the Dart side, is it invoked from FlutterView#onTouchEvent

File: shell/platform/android/io/flutter/embedding/FlutterView.java [hunk 6]
```diff
@@ -38,1015 +33,781 @@
...
-import io.flutter.embedding.legacy.AccessibilityBridge;
-import io.flutter.view.FlutterMain;
-import io.flutter.embedding.legacy.FlutterNativeView;
-import io.flutter.view.FlutterRunArguments;
-import io.flutter.view.TextureRegistry;
 import io.flutter.view.VsyncWaiter;
 
 /**
- * An Android view containing a Flutter app.
+ * {@code View} which can render a Flutter UI by attaching itself to a {@link FlutterRenderer}.
```
Comment (amirh) [inline shell/platform/android/io/flutter/embedding/FlutterView.java]: nit: Maybe worth mentioning that this is not only responsible for output (rendering) but also for input (touch events)

File: shell/platform/android/io/flutter/embedding/FlutterView.java [hunk 7]
```diff
@@ -38,1015 +33,781 @@
...
+    int physicalViewInsetLeft = 0;
+  }
+
+  private final InputMethodManager mImm;
+  private final ViewportMetrics mMetrics;
+  private final AccessibilityManager mAccessibilityManager;
+  private final AnimationScaleObserver mAnimationScaleObserver;
+
+  private FlutterRenderer flutterRenderer;
+  private BinaryMessenger pluginMessenger;
```
Comment (amirh) [inline shell/platform/android/io/flutter/embedding/FlutterView.java]: nit: I believe the messenger not only used by plugins but also by "system channels" e.g the accessibility bridge.

File: shell/platform/android/io/flutter/embedding/FlutterView.java [hunk 8]
```diff
@@ -38,1015 +33,781 @@
...
+
+  private TextInputPlugin mTextInputPlugin;
+  private MethodChannel mFlutterLocalizationChannel;
+  private BasicMessageChannel<Object> mFlutterKeyEventChannel;
+  private BasicMessageChannel<Object> mFlutterSettingsChannel;
+  private InputConnection mLastInputConnection;
+
+  private boolean isAttachedToRenderer = false;
```
Comment (amirh) [inline shell/platform/android/io/flutter/embedding/FlutterView.java]: same here: be consistent regarding m prefix for members.

File: shell/platform/android/io/flutter/embedding/FlutterView.java [hunk 9]
```diff
@@ -38,1015 +33,781 @@
...
+  }
+  //------- END ACCESSIBILITY ----
+
+  //----- START FLUTTER INTEGRATION -----
+
+  /**
+   * Start rendering the UI for the given {@link FlutterRenderer}.
+   *
+   * @param flutterRenderer the FlutterRenderer for which this FlutterView will be a RenderSurface
+   * @param pluginMessenger the BinaryMessenger to use with any plugins that this FlutterView needs
```
Comment (amirh) [inline shell/platform/android/io/flutter/embedding/FlutterView.java]: same here: I'm don't believe we usually refer to system channels as plugins

File: shell/platform/android/io/flutter/embedding/FlutterView.java [hunk 10]
```diff
@@ -38,1015 +33,781 @@
...
+   *                        to register
+   */
+  public void attachToFlutterRenderer(@NonNull FlutterRenderer flutterRenderer, @NonNull BinaryMessenger pluginMessenger) {
```
Comment (amirh) [inline shell/platform/android/io/flutter/embedding/FlutterView.java]: I'm curious, when we're running dart code in the background do we still need to be attached to a renderer? (is it guaranteed that a FlutterView exist at all?) if not, what happens with message channels?

File: shell/platform/android/io/flutter/embedding/legacy/PlatformViewsController.java [hunk 11]
```diff
@@ -50,20 +53,21 @@ public PlatformViewsController() {
...
 
-    public void attachFlutterView(FlutterView view) {
-        if (mFlutterView != null)
+    public void attachFlutterEngine(FlutterEngine flutterEngine, Context context) {
+        if (this.flutterEngine != null)
             throw new AssertionError(
-                    "A PlatformViewsController can only be attached to a single FlutterView.\n" +
-                    "attachFlutterView was called while a FlutterView was already attached."
+                    "A PlatformViewsController can only be attached to a single FlutterEngine.\n" +
+                    "attachFlutterEngine was called while a FlutterView was already attached."
```
Comment (amirh) [inline shell/platform/android/io/flutter/embedding/legacy/PlatformViewsController.java]: nit: "while a FlutterEngine was already attached"

File: shell/platform/android/io/flutter/embedding/legacy/PlatformViewsController.java [hunk 12]
```diff
@@ -41,7 +43,8 @@
 
     private final PlatformViewRegistryImpl mRegistry;
 
-    private FlutterView mFlutterView;
+    private FlutterEngine flutterEngine;
```
Comment (amirh) [inline shell/platform/android/io/flutter/embedding/legacy/PlatformViewsController.java]: same here, be consistent regarding the m prefix

File: shell/platform/android/io/flutter/embedding/legacy/PluginRegistry.java [hunk 13]
```diff
@@ -107,13 +108,6 @@
          */
         PlatformViewRegistry platformViewRegistry();
 
-        /**
-         * Returns the {@link FlutterView} that's instantiated by this plugin's
-         * {@link #activity() activity}.
-         */
-        FlutterView view();
```
Comment (amirh) [inline shell/platform/android/io/flutter/embedding/legacy/PluginRegistry.java]: Is there no reason for plugins to need the FlutterView anymore? I would assume this is going to break some plugins...
(I realize there might not be a FlutterView, maybe we can return an Optional, or a null if there's no view?)

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+ *                    the native relationships and also the PlatformViewsController relationship.
+ */
+public class FlutterEngine implements BinaryMessenger {
+  private static final String TAG = "FlutterEngine";
+
+  private final Resources resources;
+  private final FlutterJNI flutterJNI;
+  private final FlutterRenderer renderer;
+  private final FlutterPluginRegistry pluginRegistry;
+  private final Map<String, BinaryMessageHandler> mMessageHandlers;
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: Yeah, I'm systematically getting rid of the hungarian notation.  I've been thinking about composing a style doc for the Android embedding.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+  private final Map<Integer, BinaryReply> mPendingReplies = new HashMap<>();
+  private long nativeObjectReference;
+  private boolean isBackgroundView; // TODO(mattcarroll): rename to something without "view"
+  private boolean applicationIsRunning;
+  private int mNextReplyId = 1;
+
+  FlutterEngine(
+      Context context,
+      Resources resources,
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: Yes, however, what I'm doing is decomposing the individual dependencies so that I can eventually eliminate the really bad ones like Context.  For example, if all we actually needed was a Resources, we shouldn't be passing in a Context to get it.  I've noticed in many places that we're being very loose with dependencies on Context and we seem to invite that dependency just to grab smaller stuff out of it.

In other words, this is a temporary half-measure that is intended to help eliminate dependencies as I continue refactoring.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+  private final FlutterPluginRegistry pluginRegistry;
+  private final Map<String, BinaryMessageHandler> mMessageHandlers;
+  private final Map<Integer, BinaryReply> mPendingReplies = new HashMap<>();
+  private long nativeObjectReference;
+  private boolean isBackgroundView; // TODO(mattcarroll): rename to something without "view"
+  private boolean applicationIsRunning;
+  private int mNextReplyId = 1;
+
+  FlutterEngine(
+      Context context,
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: We're bringing it in specifically for the PluginRegistry, so the answer is in there somewhere.  I'm guessing it's actually cast to an Activity.  I will be re-architecting the PluginRegistry so that it does not require an Activity, and when I do that I will probably be able to eliminate the Context from this constructor all together.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+ * The {@link #runFromBundle(FlutterRunArguments)} method may not be invoked twice on the same
+ * {@code FlutterEngine}.
+ *
+ * To start rendering Flutter content to the screen, use {@link #getRenderer()} to obtain a
+ * {@link FlutterRenderer} and then attach a {@link FlutterRenderer.RenderSurface}.  Consider using
+ * a {@link FlutterView} as a {@link FlutterRenderer.RenderSurface}.
+ * TODO(mattcarroll): for the above instructions for RenderSurface to be true, we need to refactor
+ *                    the native relationships and also the PlatformViewsController relationship.
+ */
+public class FlutterEngine implements BinaryMessenger {
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: Absolutely.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+    pluginRegistry = new FlutterPluginRegistry(this, context);
+    mMessageHandlers = new HashMap<>();
+
+    attach();
+    // TODO(mattcarroll): FlutterRenderer is temporally coupled to attach(). Remove that coupling if possible.
+    this.renderer = new FlutterRenderer(flutterJNI, nativeObjectReference);
+  }
+
+  private void attach() {
+    // TODO(mattcarroll): what impact does "isBackgroundView' have?
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: Yeah, I'm equally unclear.  I think maybe this requires a conversation with @bkonyi

File: shell/platform/android/io/flutter/embedding/legacy/PluginRegistry.java (hunk 13 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/legacy/PluginRegistry.java]: I think it would help to put together a short list of very diverse plugin behaviors that could help demonstrate requirements.  Would you mind helping with that?  For example, you had a good point earlier about the video plugin.

Whatever the use-cases, I think we need to introduce "Android View attachment" as part of the plugin lifecycle.  I don't think providing a reference via "view()" is sufficient, we should give plugins an opportunity to setup and teardown when the Android View comes and goes.  Moreover, if we introduce lifecycle callbacks then we can pass the FlutterView reference in those callbacks instead of giving the impression that it is always available via "view()".

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (bkonyi) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: `isBackgroundView` determined whether or not a `FlutterNativeView` had a rendering surface specified and changed the number of threads associated with the view based on this.

FYI, `FlutterNativeView` is used directly in plugin code so this is likely to break background execution. We'll need to ensure we don't break the `android_alarm_manager` plugin again, since that breakage went unnoticed for quite awhile and was a major pain point for external devs relying on that functionality.

[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: @bkonyi Do you happen to have any ideas as to how we could reasonably get this embedding code under test?  I've been asking everyone I can think of and so far no one really has any ideas.  This code is definitely approaching an intersection with many other responsibilities and it definitely needs tests to prevent regressions.

The problem for me is that I'm used to leveraging test infrastructure that is immediately available via Gradle and Android Studio, but this embedding code is sitting in a non-project-structure, built by .gn files...

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (amirh) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: If this outlives an activity, isn't it risky to keep a reference to an Activity context...?

File: shell/platform/android/io/flutter/embedding/FlutterView.java (hunk 7 above)
Comment (amirh) [reply shell/platform/android/io/flutter/embedding/FlutterView.java]: I guess it's a terminology thing, I personally don't think of things like the accessibility bridge or things like 'flutter/platform' as plugins (plugins sounds to me like something you can optionally include in your app)

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (amirh) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: I think the last time we discussed this we figured we should probably follow the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html) as it is enforceable.
The main objection was to not kill all git history (because we need to rename all the m prefixed members), but since you are killing all history anyway maybe nos is a good opportunity.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: Sure is.  This is probably already a problem today.  As is the case with a number of other details, this is in a temporarily dangerous state until I can extricate Activitys from plugins entirely.  Because that's the real problem.  Activitys referenced by plugins is like a never ending fountain of eventual crashes.

File: shell/platform/android/io/flutter/embedding/FlutterView.java (hunk 7 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterView.java]: Interesting.  Maybe worth a discussion with some relevant people.  I thought part of Flutter's official philosophy was that "everything is a plugin."  I thought Flutter made the point publicly that "even core Flutter behavior in Android and iOS are actually just plugins."

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: @amirh If those changes are going to happen in this refactor then I'd like to re-open that conversation so I can understand the rationale for the choice.  I mean on the one hand it sounds obvious that a Google team would use the Google Java Style Guide, but also on the other hand I think Google can introduce rules that are overly restrictive and also tied to historical issues that no longer exist.  Do you know the right group of Flutter devs for us to have this conversation again?

[the remark this answers is not in the document]
Comment (Hixie) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: at least jason, amir, chris, and maurice, but it should be open to anyone else on the team with an interest.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+    return pluginRegistry;
+  }
+
+  // TODO(mattcarroll): Is this method really all about plugins? or does it have other implications?
+  public void attachViewAndActivity(Activity activity) {
+    pluginRegistry.attach(this, activity);
+  }
+
+  // TODO(mattcarroll): This method says it's unused. Is that true? Do we need it?
+  public String getObservatoryUri() {
```
Comment (Hixie) [inline shell/platform/android/io/flutter/embedding/FlutterEngine.java]: if you keep it, please make sure you use the term "Url" rathen than the obsolete term "Uri" in APIs unless there's a strong reason not to (e.g. because the type is Uri, as sometimes happens in Dart).

**Comment by Hixie** [comment]
@bparrishMines may also wish to review this

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (bkonyi) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: Honestly, I'm not totally sure what would be the best way to test this. The only method I can think of is creating a test which spawns a background isolate which is then used to send a message over a `SendPort` to the main isolate to confirm that the creation of the background isolate was successful. However, this is more of a sanity check than anything.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java (hunk 19 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: Changed the public Java API to "Url", leaving the JNI signature to avoid breaking the existing embedding.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+/**
+ * A single Flutter execution environment.
+ *
+ * A {@code FlutterEngine} can execute in the background, or it can be rendered to the screen by
+ * using the accompanying {@link FlutterRenderer}.  Rendering can be started and stopped, thus
+ * allowing a {@code FlutterEngine} to move from UI interaction to data-only processing and then
+ * back to UI interaction.
+ *
+ * To start running Flutter within this {@code FlutterEngine}, use {@link #runFromBundle(FlutterRunArguments)}.
+ * The {@link #runFromBundle(FlutterRunArguments)} method may not be invoked twice on the same
```
Comment (dnfield) [inline shell/platform/android/io/flutter/embedding/FlutterEngine.java]: may not -> must not

File: shell/platform/android/io/flutter/embedding/legacy/PluginRegistry.java (hunk 13 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/legacy/PluginRegistry.java]: I'm going to mark this conversation resolved for this PR because it will be dealt with in future PRs and requires deeper discussion.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java (hunk 20 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: done

File: shell/platform/android/io/flutter/embedding/legacy/PlatformViewsController.java (hunk 11 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/legacy/PlatformViewsController.java]: fixed

File: shell/platform/android/io/flutter/embedding/FlutterView.java (hunk 10 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterView.java]: Some of those details are yet to be specified, but I intend for background execution to not require a FlutterView at all, because to do would make little sense.

If I'm not mistake, message channels are registered with a BinaryMessenger which is implemented by FlutterEngine, not FlutterView.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: @bkonyi I've created a sample app that integrates almost all of our plugins.  Most of them work fine.  I haven't tried alarm manager.  I think it might be helpful for us to sit down together sometime this week and test some background execution.  I'm not up to date with the form that takes on Android, but I'm guessing some changes will be necessary to support the new division of responsibilities within the embedding.

File: shell/platform/android/io/flutter/embedding/FlutterEngine.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterEngine.java]: I created an empty doc that I will fill with some thoughts in parallel to this refactor and then have it reviewed by everyone:
https://docs.google.com/document/d/162Z7yn0hbttn8NSOp-vd6DFrDqpa4qhbNc2X4O0Ar3E/edit?usp=sharing

File: shell/platform/android/io/flutter/embedding/FlutterRenderer.java (hunk 3 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterRenderer.java]: @mehmetf any idea if there are docs about that lifecycle?  I'd like to comment here with the technical specifics of what it means for a surface to be created, changed, or destroyed...

File: shell/platform/android/io/flutter/embedding/FlutterView.java (hunk 9 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/FlutterView.java]: I'll follow up with this topic and make changes as necessary.  I'd like to keep the PR progress moving.

**Comment by matthew-carroll** [comment]
Assuming the build is green tomorrow, I'm going to merge this PR into the staging branch.  There don't appear to be any outstanding concerns about specifics within this PR.  There are some open issues that will need to be tackled moving forward, but I want to keep the momentum up.

I currently have a sample app that integrates most of our plugins.  I'm having some issues with firebase and video, and I haven't been able to test alarm manager yet.  But other than that things look fine.  And the issues with firebase and video are not obviously related to the refactor.  I'll work it out.
